[events-api] Create cfg file in parallel without data truncation

black-dragon74 opened the pull request:
to prevent race cases that might occur when several nodes are trying to create the config file at the same time if it is non existent (first run).

Fixes: #3714
Updates: #3715

Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 committed:
[events-api] Write cfg file with exclusive locks

to prevent race cases that might occur when several nodes
are trying to create the config file at the same time
if it is non existent (first run).

Fixes: #3714
Updates: #3715

Signed-off-by: black-dragon74 <niryadav@redhat.com>

xhernandez reviewed (commented):
What's the benefit of this approach compared to creating a temporary file and then renaming it ?

If I'm not missing anything, in case of contention, the result will be the same: all contending writers will, sooner or later, update the file. So all of them will do the same amount of work using either open + rename or LockedOpen. The only difference I see is that LockedOpen serializes those writes, while open + rename can work in parallel.

Isn't this approach slower ? is there any other benefit I missed ?

black-dragon74 commented:
> What's the benefit of this approach compared to creating a temporary file and then renaming it ?

The issue was due to truncation of the contents of the file. By serializing the writes we can ensure that other processes will write to the file only after the previous one is done with it. Creating a temp file and then renaming is not beneficial here as the file name, hence the file being written to remains the same. 

> all contending writers will, sooner or later, update the file.

Correct. But in this way of doing things we ensure there are no overlaps and hence no truncation or corruption of the data. Updating the file `n` number of times is not an issue, writing to it in parallel is.

> Isn't this approach slower ?

As we are using POSIX locks, yes this is a tad bit slower but then again we are not waiting on any external sources for the data to be written and are not writing to a temp file and then renaming it (saving one additional system call), the performance should not take a major hit.

> is there any other benefit I missed ?

None other than the base one that I can think of.

Regards

xhernandez commented:
> > What's the benefit of this approach compared to creating a temporary file and then renaming it ?
> 
> The issue was due to truncation of the contents of the file. By serializing the writes we can ensure that other processes will write to the file only after the previous one is done with it. Creating a temp file and then renaming is not beneficial here as the file name, hence the file being written to remains the same.

I don't understand this. If a temporary file name is created and then renamed to the final name, corruption is not possible. Obviously it will overwrite any previous existing file, but this also happens with the LockedOpen approach.

> 
> > all contending writers will, sooner or later, update the file.
> 
> Correct. But in this way of doing things we ensure there are no overlaps and hence no truncation or corruption of the data. Updating the file `n` number of times is not an issue, writing to it in parallel is.

There are no overlaps. Each writer will write to a different file, so even if they write in parallel, it's not a problem. Only when writes are finished, it will rename that file to the final name. Renames are atomic, so there's no risk of corruption.

> 
> > Isn't this approach slower ?
> 
> As we are using POSIX locks, yes this is a tad bit slower but then again we are not waiting on any external sources for the data to be written and are not writing to a temp file and then renaming it (saving one additional system call), the performance should not take a major hit.

In what external source may you be waiting with the open + rename approach ?

Also, you are not saving any system call. 1 open + 1 close + 1 rename are only 3 system calls. LockedOpen hides at least 2 opens + 2 closes + 1 lock (1 open, 1 close and 1 lock are inside a loop that could be executed an unbounded number of times).

So I don't see any clear benefit of the LockedOpen approach. It requires more system calls, it adds more complexity (even if it's hidden inside a class), and the final result is exactly the same. In fact I still see the open + rename approach simpler and faster.

The only thing that LockedOpen has that I think is a real benefit is that it encapsulates the logic into a reusable class. We could do the same with open + rename to make it easier to use everywhere.

black-dragon74 committed:
[events-api] Use tempfiles for parallel writes without contention

The tempfile created is different everytime, the logic of which is
encapsulated in `NamedTempOpen` class.

Once we are done writing the data to the file it is then persisted to the
disk with `os.rename()` call under the name of `filename` passed as an arg
when instantiating the class.

This approach saves us the need of fcntl locks and still guarantees the sanity
of the file.

Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 commented:
I think this should now be good to go.

/cc: @xhernandez 

Regards

black-dragon74 commented:
/recheck smoke

black-dragon74 committed:
[events-api] Close FD before renaming the file

As after rename fop, the already open file descriptor
will be stale and in a bad state which is erroneous.

Signed-off-by: black-dragon74 <niryadav@redhat.com>

xhernandez commented on events/src/utils.py:
```diff
@@ -313,6 +314,40 @@ def plugin_webhook(message):
+        This class is used to avoid the data loss or truncation in case of multiple processes
+        writing to the same file without the use of fcntl locks.
+
+        The temporary file is created in the cwd of the process.
```
I think it would be better to create it in the same directory as the file we are creating. The cwd may be anything, and in some cases it could be inappropriate to create files there.

We could even create the temporary file as hidden by prefixing it with a `.`.

black-dragon74 replied:
Noted. Addressing this as we speak.

xhernandez replied:
You are still creating the temporary file in the working directory. I think this is not correct.

black-dragon74 replied:
Nope. Please see the changes in the most recent commit. This is an outdated one. The comment is also updated wrt the change.

xhernandez commented on events/src/utils.py:
```diff
@@ -313,6 +314,40 @@ def plugin_webhook(message):
+                                   *self.open_args,
+                                   **self.open_kwargs)
+
+        if tfile is None:
```
Wouldn't it raise an exception instead of returning `None` ?

black-dragon74 replied:
It is supposed to raise an exception in case things don't go in accordance. Returning `None` would further complicate things as we only ever use this class with `with` keyword.

xhernandez replied:
If it raises exceptions in case of error, if everything goes well, it will always return an object. Why do we need to check if it's `None` ?

black-dragon74 replied:
Removed the check. Wrapped the temp file creation inside `try/except` blocks to raise the base exception.

xhernandez commented on events/src/utils.py:
```diff
@@ -313,6 +314,40 @@ def plugin_webhook(message):
+                                   **self.open_kwargs)
+
+        if tfile is None:
+            raise
```
This can only be done inside a exception handler block, otherwise you need to specify the exception.

black-dragon74 replied:
Missed that. Fixed.

xhernandez commented on events/src/utils.py:
```diff
@@ -313,6 +314,40 @@ def plugin_webhook(message):
+        return self.fileobj
+
+    def __exit__(self, ex_type, ex_val, ex_tb):
+        self.fileobj.close()
+        os.rename(self.fileobj.name, self.filename)
```
It doesn't seem to be specified if the name of the file will be present after `close()`. Probably it would be safer to read the name before closing or even do the rename before.

Also, if the exit has happened as the result of an exception, we shouldn't rename the file, just delete it.

black-dragon74 replied:
If we pass `delete=False` as a parameter to `NameTemporaryFile` it won't delete the file after `close()` so we manually have to take care of that. So in this case, doing the rename after `close` is fine as the file will be present even after close (as long as we are not using os' temp dirs, which we are not).

black-dragon74 replied:
> Also, if the exit has happened as the result of an exception, we shouldn't rename the file, just delete it.

The `__exit__` block would not be called in the case of an exception. And the exception would only happen in case temp file is not created for some reason. So, no manual work needed there as the file would not exist.

xhernandez replied:
> If we pass `delete=False` as a parameter to `NameTemporaryFile` it won't delete the file after `close()` so we manually have to take care of that. So in this case, doing the rename after `close` is fine as the file will be present even after close (as long as we are not using os' temp dirs, which we are not).

I'm not saying that the file won't be there. What I'm saying is that the `name` field of the object may be undefined after closing the file. I haven't seen any clear documentation about that, but one example in python documentation is using the name after closing the file, so I think it should be ok.

> The `__exit__` block would not be called in the case of an exception. And the exception would only happen in case temp file is not created for some reason. So, no manual work needed there as the file would not exist.

`__exit__` won't be called in case of exception inside the `__enter__` method, but it will be called if something fails inside `with` block. An exception there means that the file may be corrupt or incomplete. Renaming it to the "good" name and making it visible to everyone doesn't seem a good idea.

black-dragon74 replied:
> Renaming it to the "good" name and making it visible to everyone doesn't seem a good idea.

Agreed. Leaving the temp file as is was also not a good idea, addressed that too. If an exception occurs inside `with`, the temp file is removed from the disk.

black-dragon74 replied:
Side note: Thank you for scrutinizing the patches. Got so many new things to learn :)

black-dragon74 committed:
[events-api] Hidden temp file and use dest dir as wd

The temp file created by `NamedTempOpen` is now hidden (prefixed
with a '.'). Additionally the temp file is now created in the dest
dir of the named file, i.e. in case the filepath is /foo/bar, the
temp file will be created inside /foo.

In case the dest dir cannot be determined (cases where the path is just
a filename eg. 'myfile.json`), cwd of the process is used. However, we
will never encounter this use case as we always provide the full path.

Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 committed:
[events-api] Remove zombie file in case of exception

Signed-off-by: black-dragon74 <niryadav@redhat.com>

xhernandez commented on events/src/utils.py:
```diff
@@ -313,6 +314,56 @@ def plugin_webhook(message):
+        self.file_persisted = True
+
+    def __enter__(self):
+        try:
+            tfile = NamedTemporaryFile(mode=self.open_mode,
+                                       delete=False,
+                                       prefix='.',
+                                       dir=self.working_dir,
+                                       *self.open_args,
+                                       **self.open_kwargs)
+
+        except BaseException:
+            raise
```
```suggestion
        tfile = NamedTemporaryFile(mode=self.open_mode,
                                   delete=False,
                                   prefix='.',
                                   dir=self.working_dir,
                                   *self.open_args,
                                   **self.open_kwargs)
```

xhernandez commented on events/src/utils.py:
```diff
@@ -313,6 +314,56 @@ def plugin_webhook(message):
+        if os.path.dirname(self.filename) != "":
+            self.working_dir = os.path.dirname(self.filename)
+
+    def persist(self):
+        os.rename(self.fileobj.name, self.filename)
+        self.file_persisted = True
```
This is not needed. It can be done inside `__exit__()`. See comment in `__exit__()`.

xhernandez commented on events/src/utils.py:
```diff
@@ -313,6 +314,56 @@ def plugin_webhook(message):
+            raise
+
+        # Monkey patch the persist function onto created temp file object
+        tfile.persist = self.persist
```
Ugly...

xhernandez commented on events/src/utils.py:
```diff
@@ -313,6 +314,56 @@ def plugin_webhook(message):
+    def __exit__(self, ex_type, ex_val, ex_tb):
+        self.fileobj.close()
+
+        if not self.file_persisted:
```
Instead of that, we can simply check one of the arguments, for example `ex_type`. If it's `None`, then the `with` block completed successfully, so we can rename the file. If it's not `None`, it means that an uncaught exception has been raised. In this case we can delete the file.

black-dragon74 replied:
Got it.

xhernandez commented on events/src/utils.py:
```diff
@@ -313,6 +314,56 @@ def plugin_webhook(message):
+        self.fileobj = None
+        self.file_persisted = False
+
+        if os.path.dirname(self.filename) != "":
+            self.working_dir = os.path.dirname(self.filename)
```
nit:
```suggestion
        self.working_dir = os.path.dirname(ps.path.realpath(self.filename))
```

black-dragon74 committed:
[events-api] Refactor code as per reviews

Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

0 test(s) generated core 


2 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/00-geo-rep/00-georep-verify-non-root-setup.t
https://build.gluster.org/job/gh_centos7-regression/2973/

xhernandez commented:
@black-dragon74 this patch was accepted and only needed to pass the regression. I think this PR can be useful.

black-dragon74 commented:
> @black-dragon74 this patch was accepted and only needed to pass the regression. I think this PR can be useful.

Sorry, was doing some cleanup on my fork and this patch missed my eyes. Will open a new one.

Regards

black-dragon74 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

0 test(s) generated core 


3 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
https://build.gluster.org/job/gh_centos7-regression/3089/

black-dragon74 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

0 test(s) generated core 


4 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/features_copy-file-range.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/features_copy-file-range.t
https://build.gluster.org/job/gh_centos7-regression/3182/

black-dragon74 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

0 test(s) generated core 


4 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/features_copy-file-range.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/features_copy-file-range.t
https://build.gluster.org/job/gh_centos7-regression/3195/

xhernandez commented:
/run regression

black-dragon74 commented:
This PR is now tracked at: https://github.com/gluster/glusterfs/pull/4079 and is archived to preserve discussions.

Regards

gluster-ant commented:
!!! Couldn't read commit file !!!
